docs/guides/end-to-end-testing/testing-strategies/azure-active-directory-authentication.mdx

```diff
@@ -0,0 +1,242 @@
+limits each window to 100 requests per 15 minutes. Under test, your application
+will likely exceed this as you will be reloading your application and making
+requests to broker/verify authentication. For this demo, you can either
+
+- remove the rate limiting code (recommended for demo purposes only).
+- Implement a strategy to increase the rate limit under test.
+
+When finished, your SPA should be running on `http://localhost:3000` and be
+properly configured to run with Azure Active Directory.
+
+## Configuring Cypress to use Microsoft AAD
+
+To have access to test user credentials within our tests, we need to configure
+Cypress to use the
+[Azure Active Directory](https://learn.microsoft.com/en-us/azure/active-directory/fundamentals/active-directory-whatis)
+user credentials for a given user in your tenant. These user credentials should
+be set in the `cypress.env.json` file inside the root directory of your project.
+
+```json
+// cypress.env.json
+{
+  "aad_username": "AAD_USERNAME",
+  "aad_password": "AAD_PASSWORD",
+  "aad_name": "AAD_NAME"
+}
+```
+
+Also, to authenticate with Microsoft or Azure Active Directory, you will need to
+enable the
+[`experimentalModifyObstructiveThirdPartyCode`](/guides/guides/web-security#Modifying-Obstructive-Third-Party-Code)
+configuration option. If not enabled, your authentication workflow will enter an
+infinite redirect loop.
+
+:::cypress-config-example
+
+```js
+{
+  e2e: {
+    experimentalModifyObstructiveThirdPartyCode: true
+  }
+}
+```
+
+:::
+
+## Login with [`cy.origin()`](/api/commands/origin)
+
+Next, we'll write a custom command called `loginToAAD` to perform a login to
+[Azure Active Directory](https://learn.microsoft.com/en-us/azure/active-directory/fundamentals/active-directory-whatis).
+This command will use [`cy.origin()`](/api/commands/origin) to
+
+1. Navigate to the Azure Active Directory login page on
+   `login.microsoftonline.com`
+2. Input user credentials
+3. Sign in and redirect back to the demo application
+4. Cache the results with [`cy.session()`](/api/commands/session)
+
+```js
+// cypress/support/e2e.ts
+
+function loginViaAAD(username: string, password: string) {
+  cy.visit('http://localhost:3000/')
+  cy.get('button#signIn').click()
+
+  // Login to your AAD tenant.
+  cy.origin(
+    'login.microsoftonline.com',
+    {
+      args: {
+        username,
+      },
+    },
+    ({ username }) => {
+      cy.get('input[type="email"]').type(username, {
+        log: false,
+      })
+      cy.get('input[type="submit"]').click()
+    }
+  )
+
+  // depending on the user and how they are registered with Microsoft, the origin may go to live.com
+  cy.origin(
+    'login.live.com',
+    {
+      args: {
+        password,
+      },
+    },
+    ({ password }) => {
+      cy.get('input[type="password"]').type(password, {
+        log: false,
+      })
+      cy.get('input[type="submit"]').click()
+      cy.get('#idBtn_Back').click()
```

Comment:
whats this button? 👀

Reply:
It's the no button for "Stay signed in?". Doesn't have too much of an effect here its mainly to complete the login flow
![](https://user-images.githubusercontent.com/3980464/212986116-903c38d9-b267-4951-9893-5c8e3b76abb6.png)